File: frontend/src/utils/geo.ts
import type {Stop, StopTime} from "@/types/tranzy.ts"

const earthRadiusMeters = 6_371_000.0

export const formatMeters = (m: number): string =>
  m < 1000 ? `${Math.round(m)} m` : `${(m / 1000).toFixed(1)} km`

export const sortByDistance = <T>(
  items: T[],
  userLat: number,
  userLon: number,
  getLat: (item: T) => number,
  getLon: (item: T) => number,
  maxDistanceMeters?: number,
): T[] => {
  let mapped = items.map(item => ({item, dist: haversineMeters(userLat, userLon, getLat(item), getLon(item))}))
  if (maxDistanceMeters !== undefined) {
    mapped = mapped.filter(d => d.dist <= maxDistanceMeters)
  }
  return mapped.sort((a, b) => a.dist - b.dist).map(({item}) => item)
}

export const haversineMeters = (lat1: number, lon1: number, lat2: number, lon2: number): number => {
  const dLat = (lat2 - lat1) * (Math.PI / 180)
  const dLon = (lon2 - lon1) * (Math.PI / 180)
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * (Math.PI / 180)) * Math.cos(lat2 * (Math.PI / 180)) * Math.sin(dLon / 2) * Math.sin(dLon / 2)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
  return earthRadiusMeters * c
}

export const calculateBearing = (lat1: number, lon1: number, lat2: number, lon2: number) => {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const toDeg = (rad: number) => (rad * 180) / Math.PI;

  const phi1 = toRad(lat1);
  const phi2 = toRad(lat2);
  const deltaLambda = toRad(lon2 - lon1);

  const y = Math.sin(deltaLambda) * Math.cos(phi2);
  const x = Math.cos(phi1) * Math.sin(phi2) - Math.sin(phi1) * Math.cos(phi2) * Math.cos(deltaLambda);

  const theta = Math.atan2(y, x);
  return (toDeg(theta) + 360) % 360;
};

// Decodes a Google-encoded polyline into [lat, lng] pairs
export const decodePolyline = (encoded: string): [number, number][] => {
  const coords: [number, number][] = []
  let index = 0, lat = 0, lng = 0
  while (index < encoded.length) {
    let shift = 0, result = 0, b: number
    do { b = encoded.charCodeAt(index++) - 63; result |= (b & 0x1f) << shift; shift += 5 } while (b >= 0x20)
    lat += (result & 1) ? ~(result >> 1) : result >> 1
    shift = 0; result = 0
    do { b = encoded.charCodeAt(index++) - 63; result |= (b & 0x1f) << shift; shift += 5 } while (b >= 0x20)
    lng += (result & 1) ? ~(result >> 1) : result >> 1
    coords.push([lat / 1e5, lng / 1e5])
  }
  return coords
}

export const closestStop = (latitude: number, longitude: number, stops: StopTime[] | Stop[]): StopTime | Stop | null => {
  if (!Array.isArray(stops) || stops.length === 0) return null

  let clStop = stops[0]!
  let closestDistance = haversineMeters(latitude, longitude, clStop.stop_lat, clStop.stop_lon)

  for (let i = 1; i < stops.length; i++) {
    const currentStop = stops[i]!
    const distance = haversineMeters(latitude, longitude, currentStop.stop_lat, currentStop.stop_lon)
    if (distance < closestDistance) {
      closestDistance = distance
      clStop = currentStop
    }
  }

  return clStop
}

File: backend/handlers/geo.go
package handlers

import (
	"conexiuni-cluj/models"
	"math"
)

const (
	earthRadiusMeters  = 6_371_000.0
	averageBusSpeedMps = 25.0 / 3.6 // 25 km/h in m/s
)

func haversineMeters(lat1, lon1, lat2, lon2 float64) float64 {
	dLat := (lat2 - lat1) * math.Pi / 180
	dLon := (lon2 - lon1) * math.Pi / 180
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1*math.Pi/180)*math.Cos(lat2*math.Pi/180)*
			math.Sin(dLon/2)*math.Sin(dLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return earthRadiusMeters * c
}

func closestShapeIndex(stop models.Stop, shapes []models.Shape) int {
	return closestShapeIndexLatLon(stop.StopLat, stop.StopLon, shapes)
}

func closestShapeIndexLatLon(lat, lon float64, shapes []models.Shape) int {
	minDist := math.MaxFloat64
	idx := 0
	for i, s := range shapes {
		d := haversineMeters(lat, lon, s.ShapePtLat, s.ShapePtLon)
		if d < minDist {
			minDist = d
			idx = i
		}
	}
	return idx
}

func shapeDistanceMeters(shapes []models.Shape, fromIdx, toIdx int) float64 {
	if len(shapes) == 0 || fromIdx < 0 || toIdx < 0 || fromIdx >= toIdx || fromIdx >= len(shapes) {
		return 0
	}
	if toIdx >= len(shapes) {
		toIdx = len(shapes) - 1
	}
	var distMeters float64
	for i := fromIdx; i < toIdx; i++ {
		distMeters += haversineMeters(
			shapes[i].ShapePtLat, shapes[i].ShapePtLon,
			shapes[i+1].ShapePtLat, shapes[i+1].ShapePtLon,
		)
	}
	return distMeters
}

func calculateStopOffset(prev, curr models.Stop, shapes []models.Shape) float64 {
	fromIdx := closestShapeIndex(prev, shapes)
	toIdx := closestShapeIndex(curr, shapes)

	if fromIdx > toIdx {
		fromIdx, toIdx = toIdx, fromIdx
	}

	return shapeDistanceMeters(shapes, fromIdx, toIdx) / averageBusSpeedMps
}
